substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Target_sun_security_ssl_TrustStoreManager.java

```diff
@@ -85,36 +89,79 @@ public void afterRegistration(AfterRegistrationAccess access) {
     final Set<X509Certificate> trustedCerts;
     final KeyStore trustedKeyStore;
 
+    private static volatile Boolean truststoreSysPropSet;
+
     TrustStoreManagerSupport(Set<X509Certificate> trustedCerts, KeyStore trustedKeyStore) {
         this.trustedCerts = trustedCerts;
         this.trustedKeyStore = trustedKeyStore;
     }
+
+    static boolean useEmbeddedCerts() {
+        if (truststoreSysPropSet != null) {
+            return !truststoreSysPropSet;
+        }
+        synchronized (TrustStoreManagerSupport.class) {
+            if (truststoreSysPropSet != null) {
+                return !truststoreSysPropSet;
+            }
+            final String val = System.getProperty("javax.net.ssl.trustStore");
+            truststoreSysPropSet = val != null && !val.trim().isEmpty();
+            return !truststoreSysPropSet;
+        }
+    }
 }
 
 @TargetClass(className = TrustStoreManagerFeature.TRUST_STORE_MANAGER_CLASS_NAME)
 final class Target_sun_security_ssl_TrustStoreManager {
 
+    @Alias @RecomputeFieldValue(kind = RecomputeFieldValue.Kind.NewInstance, declClassName = TrustStoreManagerFeature.TRUST_STORE_MANAGER_CLASS_NAME +
+                    "$TrustAnchorManager") private static Target_sun_security_ssl_TrustStoreManager_TrustAnchorManager tam = null;
+
     @Substitute
     private static Set<X509Certificate> getTrustedCerts() throws Exception {
-        return ImageSingletons.lookup(TrustStoreManagerSupport.class).trustedCerts;
+        if (TrustStoreManagerSupport.useEmbeddedCerts()) {
+            return ImageSingletons.lookup(TrustStoreManagerSupport.class).trustedCerts;
+        }
+        return tam.getTrustedCerts(Target_sun_security_ssl_TrustStoreManager_TrustStoreDescriptor.createInstance());
     }
 
     @Substitute
     private static KeyStore getTrustedKeyStore() throws Exception {
-        return ImageSingletons.lookup(TrustStoreManagerSupport.class).trustedKeyStore;
+        if (TrustStoreManagerSupport.useEmbeddedCerts()) {
+            return ImageSingletons.lookup(TrustStoreManagerSupport.class).trustedKeyStore;
+        }
+        return tam.getKeyStore(Target_sun_security_ssl_TrustStoreManager_TrustStoreDescriptor.createInstance());
     }
-}
 
-/*
- * The internal classes to describe and load root certificates must not be reachable at run time.
- */
+}
 
-@Delete
 @TargetClass(className = TrustStoreManagerFeature.TRUST_STORE_MANAGER_CLASS_NAME, innerClass = "TrustStoreDescriptor")
 final class Target_sun_security_ssl_TrustStoreManager_TrustStoreDescriptor {
+
+    // we do not need these paths (corresponding to the build time environment) to be carried over
+    // to the run time
+    @Alias @RecomputeFieldValue(kind = RecomputeFieldValue.Kind.Reset) private static String defaultStorePath = null;
```

Comment:
I hope that with the proper integration of the "default store", these fields can actually be marked as `@Delete` rather than `@RecomputeFieldValue`. That would assure us that the JDK code for the default store is really properly replaced with the "store from the image heap".